src/sidebar/util/textarea-caret-position.ts

```diff
@@ -0,0 +1,91 @@
+// Hidden div used by `getCaretCoordinates` to mirror content of text area.
+let mirrorDiv: HTMLElement | undefined;
+
+// Enable debug mode. This makes the hidden div used to measure the caret
+// position visible and keeps it in the DOM after the position is calculated.
+const mirrorDebug = false;
+
+/**
+ * Calculate the coordinates of the caret position within a textarea.
+ *
+ * @return Coordinates of the caret relative to the textarea's top-left corner
+ */
+export function getCaretCoordinates(textarea: HTMLTextAreaElement): {
+  x: number;
+  y: number;
+} {
+  if (!mirrorDiv) {
+    mirrorDiv = document.createElement('div');
+    document.body.append(mirrorDiv);
+  }
+
+  try {
+    const textareaRect = textarea.getBoundingClientRect();
+
+    // Style text in the mirror to match the textarea
+    const styles = window.getComputedStyle(textarea);
+    mirrorDiv.style.font = styles.font;
+    mirrorDiv.style.padding = styles.padding;
+    mirrorDiv.style.whiteSpace = 'pre-wrap';
+    mirrorDiv.style.wordWrap = 'break-word';
+    mirrorDiv.style.position = 'fixed';
+
+    /* istanbul ignore next */
+    if (!mirrorDebug) {
+      mirrorDiv.style.visibility = 'hidden';
+    } else {
+      mirrorDiv.style.pointerEvents = 'none';
+      mirrorDiv.style.background = 'rgba(255, 0, 0, 0.5)';
+    }
+
+    // Position mirror above the textarea
+    mirrorDiv.style.width = `${textareaRect.width}px`;
```

Comment:
Perhaps you could initialize `const textareaRect = textarea.getBoundingClientRect();` right before this line, since this is the first time it's used.